crates/ibc/src/services/core/client.rs

```diff
@@ -0,0 +1,314 @@
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    ibc_context: I,
+    upgrade_context: U,
+}
+
+impl<I, U> ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyClientState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    pub fn new(ibc_context: I, upgrade_context: U) -> Self {
+        Self {
+            ibc_context,
+            upgrade_context,
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl<I, U> ClientQuery for ClientQueryService<I, U>
+where
+    I: QueryContext + Send + Sync + 'static,
+    U: UpgradeValidationContext + Send + Sync + 'static,
+    <I as ValidationContext>::AnyClientState: Into<Any>,
+    <I as ValidationContext>::AnyConsensusState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyClientState: Into<Any>,
+    <U as UpgradeValidationContext>::AnyConsensusState: Into<Any>,
+{
+    async fn client_state(
+        &self,
+        request: Request<QueryClientStateRequest>,
+    ) -> Result<Response<QueryClientStateResponse>, Status> {
+        let request_ref = request.get_ref();
+
+        trace!("Got client state request: {:?}", request_ref);
+
+        let client_id = ClientId::from_str(request_ref.client_id.as_str())?;
+        let client_state = self.ibc_context.client_state(&client_id)?;
+
+        let current_height = self.ibc_context.host_height()?;
+
+        let proof = self
+            .ibc_context
+            .get_proof(
+                current_height,
+                &Path::ClientState(ClientStatePath::new(&client_id)),
+            )
+            .ok_or_else(|| {
+                Status::not_found(format!(
+                    "Client state not found for client {} at height {}",
+                    client_id, current_height
+                ))
+            })?;
+
+        Ok(Response::new(QueryClientStateResponse {
+            client_state: Some(client_state.into()),
+            proof,
+            proof_height: Some(current_height.into()),
+        }))
+    }
+
+    async fn client_states(
+        &self,
+        request: Request<QueryClientStatesRequest>,
+    ) -> Result<Response<QueryClientStatesResponse>, Status> {
+        let request_ref = request.get_ref();
+
+        trace!("Got client states request: {:?}", request_ref);
+
+        let client_states = self.ibc_context.client_states()?;
+
+        Ok(Response::new(QueryClientStatesResponse {
+            client_states: client_states
+                .into_iter()
+                .map(|(id, state)| IdentifiedClientState {
+                    client_id: id.into(),
+                    client_state: Some(state.into()),
+                })
+                .collect(),
+            pagination: None,
```

Comment:
Would you add some comments explaining why pagination has been set to none? (for our later and users’ information as well).